aws/resource_aws_iam_service_linked_role.go

```diff
@@ -0,0 +1,169 @@
+package aws
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/iam"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceAwsIamServiceLinkedRole() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsIamServiceLinkedRoleCreate,
+		Read:   resourceAwsIamServiceLinkedRoleRead,
+		Delete: resourceAwsIamServiceLinkedRoleDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"aws_service_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: func(v interface{}, k string) (ws []string, es []error) {
+					value := v.(string)
+					if !strings.HasSuffix(value, ".amazonaws.com") {
+						es = append(es, fmt.Errorf(
+							"%q must be a service URL e.g. elasticbeanstalk.amazonaws.com", k))
+					}
+					return
+				},
+			},
+
+			"name": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"path": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"create_date": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"unique_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"description": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAwsIamServiceLinkedRoleCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).iamconn
+
+	serviceName := d.Get("aws_service_name").(string)
+
+	params := &iam.CreateServiceLinkedRoleInput{
+		AWSServiceName: aws.String(serviceName),
+	}
+
+	_, err := conn.CreateServiceLinkedRole(params)
+
+	if err != nil {
```

Comment:
I would suggest checking to see if the role already exists before creating it or handling the conflict error with something like

```go
	d.SetId(serviceName)
	if _, err := conn.CreateServiceLinkedRole(params); err != nil {
		aerr, ok := err.(awserr.Error)
		if ok && strings.Contains(aerr.Message(), "already taken") {
			if rerr := resourceAwsIamServiceLinkedRoleRead(d, meta); err != nil {
				return rerr
			}
		}
		if d.Id() == "" || !ok {
			return fmt.Errorf("Error creating service-linked role with name %s", serviceName)
		}
	}

	return nil
```

Otherwise, this will not apply cleanly on an account where the role already exist and is being used to scale production resources.

Reply:
Could we instead just import that role since you can only have one global IAM service linked role?